File: Library.HospitalSystem/Models/Patient.cs
using System.Security.Cryptography;

namespace Library.HospitalSystem.Models;

public enum Race {
    NotSpecified,
    White,
    Black,
    AmericanIndianorAlaskaNative,
    Asian,
    NativeHawaiianorOtherPacificIslander,
    Other
}

public enum Gender {
    NotSpecified,
    Male,
    Female,
    Other
}
public class Patient
{
    public uint Id { get; set; }
    public string Name { get; set; }
    public string AdressLine { get; set; }
    public DateOnly DOB { get; set; }
    public List<Race> Race { get; set; }
    public Gender Gender { get; set; }

    public Patient() {
        Id = 0;
        Name = string.Empty;
        AdressLine = string.Empty;
        DOB = DateOnly.MinValue;
        Race = new List<Race> { HospitalSystem.Models.Race.NotSpecified };
        Gender = Gender.NotSpecified;
    }

    public override string ToString()
    {
        string genderString = Gender switch
        {
            Gender.Male => "Male",
            Gender.Female => "Female",
            Gender.Other => "Other",
            _ => "UNSPECIFIED"
        };

        List<string> raceStrings = Race.Select((r) => r switch {
            Models.Race.White => "White",
            Models.Race.AmericanIndianorAlaskaNative => "American Indian or Alaska Native",
            Models.Race.Asian => "Asian",
            Models.Race.Black => "Black",
            Models.Race.NativeHawaiianorOtherPacificIslander => "Native Hawaiia or Other Pacific Islander",
            Models.Race.Other => "Other",
            _ => "UNSPECIFIED"
        }).ToList();

        return $"{Id} {Name} {AdressLine} {DOB} {genderString} {string.Join(", ", raceStrings)}";
    }
}

File: next-frontend-hospital-system/src/types/api/patient.ts
export interface PatientDTO {
    id: number;
    name: string;
    addressLine: string;
    dob: string;
    gender: string;
    race: [string, ...string[]];
    insurancePlan: InsurancePlanDto;
}

/*
public class InsurancePlanDto
{
    public int Id { get; set; } = 0;
    public string Name { get; set; } = string.Empty;
    public double Deductable { get; set; } = 0;
    public double Copay { get; set; } = 0;
    public double CoinsurancePercent { get; set; } = 0;
    public double OOPM { get; set; } = 0;
}
*/

export interface InsurancePlanDto {
    id: number;
    name: string;
    deductable: number;
    copay: number;
    coinsurancePercent: number;
    oopm: number;
}

export interface InsurancePlanCreateDto {
    name: string;
    deductable: number;
    copay: number;
    coinsurancePercent: number;
    oopm: number;
}

/*public enum Race {
    NotSpecified,
    White,
    Black,
    AmericanIndianorAlaskaNative,
    Asian,
    NativeHawaiianorOtherPacificIslander,
    Other
}

public enum Gender {
    NotSpecified,
    Male,
    Female,
    Other
}*/

export function convertRace(race: string) {
    switch (race) {
        case "notSpecified":
            return "Not Specified";
        case "white":
            return "White";
        case "black":
            return "Black";
        case "americanIndianorAlaskaNative":
            return "American-Indian or Alaska native";
        case "asian":
            return "Asian";
        case "nativeHawaiianorOtherPacificIslander":
            return "Native Hawaiian or Other Pacific Islander";
        case "other":
            return "Other";
    }
}

export function convertGender(gender: string) {
    switch (gender) {
        case "notSpecified":
            return "N/S";
        case "male":
            return "Male";
        case "female":
            return "Female";
        case "other":
            return "Other";
    }
}
